### PR TITLE
Enhance Slice API

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
Enhances Slice API:
- Provides the ability to specify anchor and shape as normalized or absolute coordinates (independently)
- Provides the ability to specify order and number of slice arguments provided by name (as in the layout) and by indexes:

Examples:
`dim_names = "WH"` : anchor and shape will follow `x, y` order for any layout (will slice dimensions 1 and 0 for `HWC` and dimensions 2 and 1 for `CHW`
`dims = (1, 0)`: anchor and shape will refer to second and first dimensions (that is, `x` and `y` for `HWC` layout, but would mean `y` and `c` for `CHW`) 

- Allows 3D slicing (depth slicing)
- Allows slicing of channel dimension
- Allows slicing of frame dimension
- Default values are normalized anchor and shape, and `dim_names="WH"`

#### What happened in this PR?
- Implemented API described above
- Wrote tests for all those options

**JIRA TASK**: [DALI-1034]

### CODE REVIEW

File: dali/test/python/test_operator_slice.py [hunk 1]
```diff
@@ -198,14 +225,14 @@ def slice_func(image):
...
-def test_slice_vs_numpy_slice_gpu():
+def tes2t_slice_vs_numpy_slice_gpu():
     for batch_size in {1, 32, 64}:
         eii = SliceArgsIteratorAllDims(batch_size)
         compare_pipelines(SlicePipeline('gpu', batch_size, iter(eii)),
                           PythonOperatorPipeline(slice_func, batch_size),
                           batch_size=batch_size, N_iterations=10)
 
-def test_slice_vs_numpy_slice_cpu():
+def te2st_slice_vs_numpy_slice_cpu():
```
Comment (jantonguirao) [inline dali/test/python/test_operator_slice.py]: revert that

File: dali/pipeline/operators/crop/slice.h [hunk 2]
```diff
@@ -20,24 +20,28 @@
...
 #include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/crop/slice_attr.h"
 #include "dali/pipeline/operators/crop/slice_base.h"
 #include "dali/pipeline/operators/operator.h"
 
 namespace dali {
 
 template <typename Backend>
-class Slice : public SliceBase<Backend> {
+class Slice : public SliceBase<Backend>, protected SliceAttr {
```
Comment (JanuszL) [inline dali/pipeline/operators/crop/slice.h]: Maybe it is a good time to change it from the inheritance to composition?

File: dali/pipeline/operators/crop/slice_attr.h [hunk 3]
```diff
@@ -19,105 +19,131 @@
...
-                data_idx,
-                crop_size.data<float>()[0],
-                crop_size.data<float>()[1],
-                crop_begin.data<float>()[0],
-                crop_begin.data<float>()[1]);
-        }
+  void ProcessArguments(DeviceWorkspace &ws) {
+    DALI_ENFORCE(ws.NumInput() == 3,
+      "Expected 3 inputs. Received: " + std::to_string(ws.NumInput()));
+    const auto &images = ws.Input<GPUBackend>(0);
```
Comment (JanuszL) [inline dali/pipeline/operators/crop/slice_attr.h]: Unused

File: dali/pipeline/operators/crop/slice_attr.h [hunk 4]
```diff
@@ -19,105 +19,131 @@
...
+      VerifyArgsShape(crop_anchor.tensor_shape(data_idx), crop_shape.tensor_shape(data_idx));
+      ProcessArgumentsHelper(data_idx, crop_anchor.tensor<float>(data_idx),
+                             crop_shape.tensor<float>(data_idx));
     }
+  }
+
+  void ProcessArguments(const SampleWorkspace &ws) {
+    DALI_ENFORCE(ws.NumInput() == 3,
+      "Expected 3 inputs. Received: " + std::to_string(ws.NumInput()));
+    const auto &images = ws.Input<CPUBackend>(0);
```
Comment (JanuszL) [inline dali/pipeline/operators/crop/slice_attr.h]: Unused

File: dali/pipeline/operators/crop/slice.h (hunk 2 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/slice.h]: done

File: dali/pipeline/operators/crop/slice_attr.h (hunk 3 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/slice_attr.h]: done

File: dali/pipeline/operators/crop/slice_attr.h (hunk 4 above)
Comment (jantonguirao) [reply dali/pipeline/operators/crop/slice_attr.h]: done

**Comment by jantonguirao** [comment]
!build

File: dali/operators/crop/slice.cc [hunk 5]
```diff
@@ -19,36 +19,36 @@ namespace dali {
...
+coordinates and `WH` order for the slice arguments.)code")
     .NumInput(3)
     .NumOutput(1)
     .AllowSequences()
     .AddOptionalArg("image_type",
       R"code(The color space of input and output image)code",
       DALI_RGB, false)
     .AddParent("SliceBase")
+    .AddParent("SliceAttr")
     .InputLayout(0, { "HW", "HWC", "DHWC" });
```
Comment (mzient) [inline dali/operators/crop/slice.cc]: I'd lift that constraint and handle layouts manually or add another flavor (e.g. DefaultInputLayout) that sets one of these when absent, but accepts anything with number of components matching the dimensionality of the input.

Comment (jantonguirao) [reply dali/operators/crop/slice.cc]: Done

**Comment by jantonguirao** [comment]
!build

File: dali/operators/crop/slice_attr.h [hunk 6]
```diff
@@ -19,105 +19,129 @@
...
+        slice.shape = shape;
 
-    DALI_ENFORCE(crop_x_norm + crop_w <= 1.0f,
-      "crop_x[" + std::to_string(crop_x_norm) + "] + crop_width["
-      + std::to_string(crop_w) + "] must be <= 1.0f");
-    DALI_ENFORCE(crop_y_norm + crop_h <= 1.0f,
-      "crop_y[" + std::to_string(crop_y_norm) + "] + crop_height["
-      + std::to_string(crop_h) + "] must be <= 1.0f");
+        if (!dim_names_.empty()) {
+          dims_ = {};
```
Comment (klecki) [inline dali/operators/crop/slice_attr.h]: Isn't that a race? On CPU all the threads in thread pool can overwrite `dims_` on their own, clearing it and pushing back to it in a mixed order.

Comment (jantonguirao) [reply dali/operators/crop/slice_attr.h]: You are right. Will fix

Comment (jantonguirao) [reply dali/operators/crop/slice_attr.h]: Done

**Comment by jantonguirao** [comment]
!build

**Pull request merged**